File: src/virtualization/bin/guest/socat.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/virtualization/bin/guest/socat.h"

#include <fuchsia/virtualization/cpp/fidl.h>
#include <lib/async-loop/cpp/loop.h>
#include <lib/async-loop/default.h>
#include <lib/syslog/cpp/macros.h>
#include <lib/zx/status.h>
#include <zircon/status.h>

#include <iostream>

#include "lib/fidl/cpp/binding.h"
#include "src/lib/fsl/handles/object_info.h"
#include "src/virtualization/bin/guest/serial.h"
#include "src/virtualization/bin/guest/services.h"

namespace {

class HostVsockAcceptor : public fuchsia::virtualization::HostVsockAcceptor {
 public:
  HostVsockAcceptor(uint32_t port, async::Loop* loop) : port_(port), console_(loop) {}

  // |fuchsia::virtualization::HostVsockAcceptor|
  void Accept(uint32_t src_cid, uint32_t src_port, uint32_t port,
              AcceptCallback callback) override {
    if (port != port_) {
      std::cerr << "Unexpected port " << port << "\n";
      callback(ZX_ERR_CONNECTION_REFUSED, zx::handle());
      return;
    }
    zx::socket socket, remote_socket;
    zx_status_t status = zx::socket::create(ZX_SOCKET_STREAM, &socket, &remote_socket);
    if (status != ZX_OK) {
      std::cerr << "Failed to create socket " << status << "\n";
      callback(ZX_ERR_CONNECTION_REFUSED, zx::handle());
      return;
    }
    callback(ZX_OK, std::move(remote_socket));
    console_.Start(zx::socket(std::move(socket)));
  }

 private:
  uint32_t port_;
  GuestConsole console_;
};

// Connect the the vsock endpoint of the given environment.
zx::status<fuchsia::virtualization::HostVsockEndpointSyncPtr> connect_to_vsock_endpoint(
    uint32_t env_id, uint32_t port, sys::ComponentContext* context) {
  // Connect to environment.
  zx::status<fuchsia::virtualization::RealmSyncPtr> realm = ConnectToEnvironment(context, env_id);
  if (realm.is_error()) {
    return realm.take_error();
  }

  fuchsia::virtualization::HostVsockEndpointSyncPtr vsock_endpoint;
  zx_status_t status = realm->GetHostVsockEndpoint(vsock_endpoint.NewRequest());
  if (status != ZX_OK) {
    std::cerr << "Could not fetch vsock endpoint: " << zx_status_get_string(status) << ".\n";
    return zx::error(status);
  }

  return zx::ok(std::move(vsock_endpoint));
}

}  // namespace

zx_status_t handle_socat_listen(uint32_t env_id, uint32_t port, async::Loop* loop,
                                sys::ComponentContext* context) {
  zx::status<fuchsia::virtualization::HostVsockEndpointSyncPtr> vsock_endpoint =
      connect_to_vsock_endpoint(env_id, port, context);
  if (vsock_endpoint.is_error()) {
    return vsock_endpoint.error_value();
  }

  HostVsockAcceptor acceptor(port, loop);
  fidl::Binding<fuchsia::virtualization::HostVsockAcceptor> binding(&acceptor);
  zx_status_t status;
  vsock_endpoint->Listen(port, binding.NewBinding(), &status);
  if (status != ZX_OK) {
    std::cerr << "Failed to listen on port " << port << "\n";
    return status;
  }

  return loop->Run();
}

zx_status_t handle_socat_connect(uint32_t env_id, uint32_t cid, uint32_t port, async::Loop* loop,
                                 sys::ComponentContext* context) {
  zx::status<fuchsia::virtualization::HostVsockEndpointSyncPtr> vsock_endpoint =
      connect_to_vsock_endpoint(env_id, port, context);
  if (vsock_endpoint.is_error()) {
    return vsock_endpoint.error_value();
  }

  zx::socket socket, remote_socket;
  zx_status_t status = zx::socket::create(ZX_SOCKET_STREAM, &socket, &remote_socket);
  if (status != ZX_OK) {
    std::cerr << "Failed to create socket " << status << "\n";
    return status;
  }
  vsock_endpoint->Connect(cid, port, std::move(remote_socket), &status);
  if (status != ZX_OK) {
    std::cerr << "Failed to connect " << status << "\n";
    return status;
  }

  GuestConsole console(loop);
  console.Start(std::move(socket));

  return loop->Run();
}

File: src/virtualization/bin/guest/src/socat.rs
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
use {
    crate::{arguments, services},
    anyhow::anyhow,
    anyhow::Context,
    anyhow::Error,
    fidl_fuchsia_virtualization::{
        GuestStatus, HostVsockAcceptorMarker, HostVsockEndpointMarker, HostVsockEndpointProxy,
    },
    fuchsia_zircon as zx, fuchsia_zircon_status as zx_status,
    futures::TryStreamExt,
};

pub async fn connect_to_vsock_endpoint(
    guest_type: arguments::GuestType,
) -> Result<HostVsockEndpointProxy, Error> {
    let guest_manager = services::connect_to_manager(guest_type)?;
    let guest_info = guest_manager.get_info().await?;
    if guest_info.guest_status.unwrap() == GuestStatus::Running {
        let guest_endpoint = services::connect(guest_type).await?;
        let (vsock_endpoint, vsock_server_end) =
            fidl::endpoints::create_proxy::<HostVsockEndpointMarker>()
                .context("failed to make vsock endpoint")?;
        guest_endpoint
            .get_host_vsock_endpoint(vsock_server_end)
            .await?
            .map_err(|err| anyhow!("failed to get HostVsockEndpoint: {:?}", err))?;
        Ok(vsock_endpoint)
    } else {
        Err(anyhow!(zx_status::Status::NOT_CONNECTED))
    }
}

pub async fn handle_socat_listen(
    vsock_endpoint: HostVsockEndpointProxy,
    host_port: u32,
) -> Result<(), Error> {
    let (vsock_accept_client, mut vsock_acceptor_stream) =
        fidl::endpoints::create_request_stream::<HostVsockAcceptorMarker>()
            .context("failed to make vsock acceptor")?;

    vsock_endpoint
        .listen(host_port, vsock_accept_client)
        .await?
        .map_err(|val| zx::Status::from_raw(val))?;

    // Try_next returns a Result<Option<T>, Err>, hence we need to unwrap our value
    let connection =
        vsock_acceptor_stream.try_next().await?.ok_or(zx_status::Status::UNAVAILABLE)?;
    let (_src_cid, _src_port, port, responder) =
        connection.into_accept().ok_or(zx_status::Status::CONNECTION_REFUSED)?;

    if port != host_port {
        responder.send(&mut Err(zx_status::Status::CONNECTION_REFUSED.into_raw()))?;
        return Err(anyhow!(zx_status::Status::CONNECTION_REFUSED));
    }

    let (socket, remote_socket) = zx::Socket::create(zx::SocketOpts::STREAM)?;
    responder.send(&mut Ok(remote_socket))?;
    let io = services::GuestConsole::new(socket)?;

    io.run_with_stdio().await.map_err(From::from)
}

pub async fn handle_socat(vsock_endpoint: HostVsockEndpointProxy, port: u32) -> Result<(), Error> {
    let socket = vsock_endpoint.connect(port).await?.map_err(|val| zx::Status::from_raw(val))?;
    let io = services::GuestConsole::new(socket)?;
    io.run_with_stdio().await.map_err(From::from)
}

#[cfg(test)]
mod test {
    use {
        super::*, assert_matches::assert_matches, fidl::endpoints::create_proxy_and_stream,
        fuchsia_async as fasync, futures::future::join, futures::StreamExt,
        pretty_assertions::assert_eq,
    };

    #[fasync::run_until_stalled(test)]
    async fn socat_listen_invalid_host_returns_err() {
        let (proxy, mut stream) = create_proxy_and_stream::<HostVsockEndpointMarker>().unwrap();
        let server = async move {
            let (port, _acceptor, responder) = stream
                .next()
                .await
                .expect("Failed to read from stream")
                .expect("Failed to parse request")
                .into_listen()
                .expect("Unexpected call to Guest Proxy");
            assert_eq!(port, 0);
            responder
                .send(&mut Err(zx_status::Status::CONNECTION_REFUSED.into_raw()))
                .expect("Failed to send status code to client");
        };

        let client = handle_socat_listen(proxy, 0);
        let (_server_res, client_res) = join(server, client).await;
        assert_matches!(
            client_res.unwrap_err().downcast(),
            Ok(zx_status::Status::CONNECTION_REFUSED)
        );
    }

    #[fasync::run_until_stalled(test)]
    async fn socat_listen_mismatched_ports_returns_err() {
        let (proxy, mut stream) = create_proxy_and_stream::<HostVsockEndpointMarker>().unwrap();
        let server = async move {
            let (port, acceptor, responder) = stream
                .next()
                .await
                .expect("Failed to read from stream")
                .expect("Failed to parse request")
                .into_listen()
                .expect("Unexpected call to Guest Proxy");
            assert_eq!(port, 0);
            responder.send(&mut Ok(())).expect("Failed to send status code to client");
            let _ = acceptor
                .into_proxy()
                .expect("Failed to convert client end into proxy")
                .accept(0, 0, 1)
                .await
                .expect("Failed to accept listener");
        };

        let client = handle_socat_listen(proxy, 0);
        let (_server_res, client_res) = join(server, client).await;
        assert_matches!(
            client_res.unwrap_err().downcast(),
            Ok(zx_status::Status::CONNECTION_REFUSED)
        );
    }
}
